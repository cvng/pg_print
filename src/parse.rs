use crate::fmt::Printer;
use crate::stmt::RawStmt;

impl Printer {
    #[cfg(feature = "unstable")]
    pub fn parse(&mut self, parse: &parser::Parse) {
        self.cbox(0);
        for stmt in &parse.stmts {
            self.stmt(&RawStmt { stmt: &stmt.stmt });
        }
        self.end();
    }

    #[cfg(not(feature = "unstable"))]
    pub fn parse(&mut self, parse: &pg_query::protobuf::ParseResult) {
        self.cbox(0);
        for stmt in &parse.stmts {
            if let Some(stmt) = &stmt.stmt {
                if let Some(node) = &stmt.node {
                    self.stmt(&RawStmt { stmt: node });
                }
            }
        }
        self.end();
    }
}
