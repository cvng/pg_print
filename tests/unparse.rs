use insta::assert_snapshot;
use pg_query::parse;
use std::env;
use std::fs;
use std::path::PathBuf;

#[test]
fn unparse() {
    let path = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("tests/unparse.sql")
        .strip_prefix(env::current_dir().unwrap())
        .unwrap()
        .to_path_buf();

    for (mut lineno, case) in fs::read_to_string(&path).unwrap().lines().enumerate() {
        lineno += 1;

        if case.starts_with("--") {
            continue;
        }

        let deparsed = pg_query::deparse(&parse(case).unwrap().protobuf)
            .unwrap()
            .to_string();

        let unparsed = pg_print::unparse(&parse(case).unwrap().protobuf)
            .unwrap()
            .to_string();

        let reparsed = pg_query::deparse(&parse(&unparsed).unwrap().protobuf)
            .unwrap()
            .to_string();

        let fingerprint = pg_query::fingerprint(case).unwrap().hex.to_string();

        assert_eq!(deparsed, reparsed);

        assert_snapshot!(
            fingerprint,
            unparsed,
            &format!("{}:{}", path.display(), lineno)
        );
    }
}
