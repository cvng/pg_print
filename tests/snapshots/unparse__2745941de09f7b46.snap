---
source: tests/unparse.rs
expression: "tests/unparse.sql:126"
---
create domain us_postal_code as text check (value ~ E'^\\d{5}$' or value ~ E'^\\d{5}-\\d{4}$')
